[Safe CPP] Address warnings in rendering/

stwrt opened the pull request:
#### e1a0394f3485c6932633210588bcc464d2b96c5f
<pre>
[Safe CPP] Address warnings in rendering/
<a href="https://bugs.webkit.org/show_bug.cgi?id=302167">https://bugs.webkit.org/show_bug.cgi?id=302167</a>
<a href="https://rdar.apple.com/problem/164270032">rdar://problem/164270032</a>

Reviewed by Chris Dumez.

Address warnings in RenderHighlight, RenderSelection, and TextBoxPainter cpp and header files.

* Source/WebCore/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations
* Source/WebCore/rendering/RenderHighlight.cpp
* Source/WebCore/rendering/RenderHighlight.h
* Source/WebCore/rendering/RenderSelection.cpp
* Source/WebCore/rendering/RenderSelection.h
* Source/WebCore/rendering/RenderSelectionGeometry.h
* Source/WebCore/rendering/TextBoxPainter.cpp
* Source/WebCore/rendering/TextBoxPainter.h

* Source/WebCore/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/rendering/RenderHighlight.cpp:
(WebCore::RenderRangeIterator::current const):
(WebCore::RenderRangeIterator::next):
(WebCore::RenderRangeIterator::checkForSpanner):
(WebCore::rendererAfterOffset):
(WebCore::RenderHighlight::setRenderRange):
(WebCore::RenderHighlight::highlightStateForRenderer):
* Source/WebCore/rendering/RenderHighlight.h:
* Source/WebCore/rendering/RenderSelection.cpp:
(WebCore::rendererAfterOffset):
(WebCore::containingBlockBelowView):
(WebCore::collectSelectionData):
(WebCore::RenderSelection::set):
(WebCore::RenderSelection::clear):
(WebCore::RenderSelection::repaint const):
(WebCore::RenderSelection::collectBounds const):
(WebCore::RenderSelection::apply):
* Source/WebCore/rendering/RenderSelection.h:
* Source/WebCore/rendering/RenderSelectionGeometry.h:
(WebCore::RenderSelectionGeometryBase::checkedRepaintContainer const):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::m_isCombinedText):
(WebCore::m_haveSelection):
(WebCore::TextBoxPainter::paint):
(WebCore::TextBoxPainter::selectionStartEnd const):
(WebCore::TextBoxPainter::paintCompositionForeground):
(WebCore::TextBoxPainter::paintForegroundAndDecorations):
(WebCore::TextBoxPainter::paintBackgroundFill):
(WebCore::TextBoxPainter::selectionRectForRange const):
(WebCore::TextBoxPainter::paintBackgroundFillForRange):
(WebCore::TextBoxPainter::paintForeground):
(WebCore::TextBoxPainter::paintForegroundForShapeRange):
(WebCore::TextBoxPainter::createDecorationPainter):
(WebCore::isDecoratingBoxForBackground):
(WebCore::TextBoxPainter::collectDecoratingBoxesForBackgroundPainting):
(WebCore::TextBoxPainter::paintBackgroundDecorations):
(WebCore::TextBoxPainter::paintForegroundDecorations):
(WebCore::TextBoxPainter::fillCompositionUnderline const):
(WebCore::TextBoxPainter::paintCompositionUnderlines):
(WebCore::TextBoxPainter::paintCompositionUnderline):
(WebCore::TextBoxPainter::paintPlatformDocumentMarker):
(WebCore::TextBoxPainter::computeHaveSelection const):
(WebCore::TextBoxPainter::fontCascade const):
(WebCore::TextBoxPainter::textOriginFromPaintRect const):
* Source/WebCore/rendering/TextBoxPainter.h:
(WebCore::TextBoxPainter::writingMode const):

Canonical link: <a href="https://commits.webkit.org/303440@main">https://commits.webkit.org/303440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d900e37facd919093c9a58933cdc04b20151b5fb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132415 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4910 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139930 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84372 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2db87e00-aebe-460c-895a-0fde2a40b224) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134285 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4928 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4669 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101227 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f8ceee16-b895-4819-9bf3-8858f1ee7b34) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135361 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118606 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82019 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83157 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112191 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36723 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142582 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4580 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37312 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109603 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3949 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109782 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27821 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3465 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114880 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57856 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4634 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33234 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4466 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68085 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4725 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4591 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/rendering/RenderHighlight.cpp:
```diff
@@ -57,35 +57,37 @@ RenderRangeIterator::RenderRangeIterator(RenderObject* start)
-    RenderObject* currentSpan = m_spannerStack.isEmpty() ? nullptr : m_spannerStack.last()->spanner();
-    m_current = m_current->nextInPreOrder(currentSpan);
+    CheckedPtr currentSpan = m_spannerStack.isEmpty() ? nullptr : m_spannerStack.last()->spanner();
+    CheckedPtr checkedSpan = currentSpan.get();
```
We don't normally use `checked` or `protected` in variable names, and we try to keep the names short.

cdumez commented on Source/WebCore/rendering/RenderHighlight.cpp:
```diff
@@ -57,35 +57,37 @@ RenderRangeIterator::RenderRangeIterator(RenderObject* start)
-    m_current = m_current->nextInPreOrder(currentSpan);
+    CheckedPtr currentSpan = m_spannerStack.isEmpty() ? nullptr : m_spannerStack.last()->spanner();
+    CheckedPtr checkedSpan = currentSpan.get();
+    CheckedPtr checkedCurrent = m_current.get();
```
ditto, I'd use `current` here.

cdumez commented on Source/WebCore/rendering/RenderHighlight.h:
```diff
@@ -70,11 +70,11 @@ class RenderRangeIterator {
-    
-    RenderObject* m_current { nullptr };
+
+    CheckedPtr<RenderObject> m_current { nullptr };
```
`{ nullptr }` is unnecessary

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -48,7 +48,7 @@ namespace WebCore {
 namespace {
 
 struct SelectionContext {
-    
```
You should update your IDE to avoid this noise in PRs. We prefer not to touch spaces on lines we don't update.

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -92,17 +92,17 @@ static SelectionContext collectSelectionData(const RenderRange& selection, bool
-        stop = rendererAfterOffset(*selection.end(), selection.endOffset());
-    RenderRangeIterator selectionIterator(start);
+    CheckedPtr start = selection.start();
+    CheckedPtr<RenderObject> stop = nullptr;
```
`= nullptr` is unnecessary.

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -143,24 +143,26 @@ void RenderSelection::set(const RenderRange& selection, RepaintMode blockRepaint
-        m_renderView.layer()->repaintBlockSelectionGaps();
+    if (!m_selectionWasCaret) {
+        CheckedPtr layer = m_renderView->layer();
+        layer->repaintBlockSelectionGaps();
```
Should really call `checkedLayer()` which already exists I think even. No point in adding a local variable.

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -143,24 +143,26 @@ void RenderSelection::set(const RenderRange& selection, RepaintMode blockRepaint
-    RenderObject* end = nullptr;
-    if (m_renderRange.end())
-        end = rendererAfterOffset(*m_renderRange.end(), m_renderRange.endOffset());
+    CheckedPtr<RenderObject> end = nullptr;
```
`= nullptr` is unnecessary

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -171,21 +173,21 @@ void RenderSelection::repaint() const
-    
-    RenderRangeIterator selectionIterator(start);
+    CheckedPtr start = m_renderRange.start();
+    CheckedPtr<RenderObject> stop = nullptr;
```
`= nullptr` is unnecessary.

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -225,102 +227,108 @@ void RenderSelection::apply(const RenderRange& newSelection, RepaintMode blockRe
-    RenderObject* selectionEnd = nullptr;
-    auto* selectionDataEnd = m_renderRange.end();
+
+    CheckedPtr<RenderObject> selectionEnd = nullptr;
```
`= nullptr` is unnecessary.

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -225,102 +227,108 @@ void RenderSelection::apply(const RenderRange& newSelection, RepaintMode blockRe
+
+    if (blockRepaintMode != RepaintMode::Nothing) {
+        CheckedPtr layer = m_renderView->layer();
+        layer->clearBlockSelectionGapsBounds();
```
Should call `checkedLayer()`

cdumez commented on Source/WebCore/rendering/RenderSelection.cpp:
```diff
@@ -225,102 +227,108 @@ void RenderSelection::apply(const RenderRange& newSelection, RepaintMode blockRe
 #if ENABLE(SERVICE_CONTROLS)
-                m_selectionGeometryGatherer.addGapRects(blockSelectionGeometry->repaintContainer(), blockSelectionGeometry->rects());
+                CheckedPtr blockRepaintContainer = blockSelectionGeometry->repaintContainer();
+                m_selectionGeometryGatherer.addGapRects(blockRepaintContainer.get(), blockSelectionGeometry->rects());
```
We should probably add a `checkedRepaintContainer()` getter on `RenderSelectionGeometryBase`

cdumez commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -176,29 +176,29 @@ static void buildTextForShaping(ShapedContent& shapedContent, InlineIterator::Bo
-    }())
-    , m_isPrinting(m_document.printing())
-    , m_haveSelection(computeHaveSelection())
+: m_textBox(InlineIterator::BoxModernPath { inlineContent, inlineContent.indexForBox(box) })
```
I think something is fishy with the indentation here no? I think the previous indention was correct unless I'm missing something.

cdumez commented on Source/WebCore/rendering/RenderHighlight.cpp:
```diff
@@ -57,35 +57,36 @@ RenderRangeIterator::RenderRangeIterator(RenderObject* start)
-    RenderObject* currentSpan = m_spannerStack.isEmpty() ? nullptr : m_spannerStack.last()->spanner();
-    m_current = m_current->nextInPreOrder(currentSpan);
+    CheckedPtr span = m_spannerStack.isEmpty() ? nullptr : m_spannerStack.last()->spanner();
+    CheckedPtr current = m_current.get();
```
We probably don't want this variable here since it is outdated right on the next line and only used once. I'd recommend using `CheckedRef { *m_current }->` not the next line instead.

stwrt replied:
Fixed

cdumez reviewed (approved):
LGTM with nit.

stwrt committed:
[Safe CPP] Address warnings in rendering/
https://bugs.webkit.org/show_bug.cgi?id=302167
rdar://problem/164270032

Reviewed by Chris Dumez.

Address warnings in RenderHighlight, RenderSelection, and TextBoxPainter cpp and header files.

* Source/WebCore/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations
* Source/WebCore/rendering/RenderHighlight.cpp
* Source/WebCore/rendering/RenderHighlight.h
* Source/WebCore/rendering/RenderSelection.cpp
* Source/WebCore/rendering/RenderSelection.h
* Source/WebCore/rendering/RenderSelectionGeometry.h
* Source/WebCore/rendering/TextBoxPainter.cpp
* Source/WebCore/rendering/TextBoxPainter.h

* Source/WebCore/SaferCPPExpectations/NoUncheckedPtrMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/rendering/RenderHighlight.cpp:
(WebCore::RenderRangeIterator::current const):
(WebCore::RenderRangeIterator::next):
(WebCore::RenderRangeIterator::checkForSpanner):
(WebCore::rendererAfterOffset):
(WebCore::RenderHighlight::setRenderRange):
(WebCore::RenderHighlight::highlightStateForRenderer):
* Source/WebCore/rendering/RenderHighlight.h:
* Source/WebCore/rendering/RenderSelection.cpp:
(WebCore::rendererAfterOffset):
(WebCore::containingBlockBelowView):
(WebCore::collectSelectionData):
(WebCore::RenderSelection::set):
(WebCore::RenderSelection::clear):
(WebCore::RenderSelection::repaint const):
(WebCore::RenderSelection::collectBounds const):
(WebCore::RenderSelection::apply):
* Source/WebCore/rendering/RenderSelection.h:
* Source/WebCore/rendering/RenderSelectionGeometry.h:
(WebCore::RenderSelectionGeometryBase::checkedRepaintContainer const):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::m_isCombinedText):
(WebCore::m_haveSelection):
(WebCore::TextBoxPainter::paint):
(WebCore::TextBoxPainter::selectionStartEnd const):
(WebCore::TextBoxPainter::paintCompositionForeground):
(WebCore::TextBoxPainter::paintForegroundAndDecorations):
(WebCore::TextBoxPainter::paintBackgroundFill):
(WebCore::TextBoxPainter::selectionRectForRange const):
(WebCore::TextBoxPainter::paintBackgroundFillForRange):
(WebCore::TextBoxPainter::paintForeground):
(WebCore::TextBoxPainter::paintForegroundForShapeRange):
(WebCore::TextBoxPainter::createDecorationPainter):
(WebCore::isDecoratingBoxForBackground):
(WebCore::TextBoxPainter::collectDecoratingBoxesForBackgroundPainting):
(WebCore::TextBoxPainter::paintBackgroundDecorations):
(WebCore::TextBoxPainter::paintForegroundDecorations):
(WebCore::TextBoxPainter::fillCompositionUnderline const):
(WebCore::TextBoxPainter::paintCompositionUnderlines):
(WebCore::TextBoxPainter::paintCompositionUnderline):
(WebCore::TextBoxPainter::paintPlatformDocumentMarker):
(WebCore::TextBoxPainter::computeHaveSelection const):
(WebCore::TextBoxPainter::fontCascade const):
(WebCore::TextBoxPainter::textOriginFromPaintRect const):
* Source/WebCore/rendering/TextBoxPainter.h:
(WebCore::TextBoxPainter::writingMode const):

Canonical link: https://commits.webkit.org/303440@main
